### ISSUE TITLE
[Bug]: ConcurrentModificationException again on 5.0.4

### ISSUE THREAD

**Issue by sarbyn** [OneSignal/OneSignal-Android-SDK#1920]
### What happened?

We got in our production build (with Sdk version 5.0.0) a lot of issues in 5 different places, that seems related to ticket #1856 

These are the stack traces

```
Fatal Exception: java.util.ConcurrentModificationException:
       at java.util.ArrayList$Itr.next(ArrayList.java:860)
       at kotlin.collections.CollectionsKt___CollectionsKt.firstOrNull(CollectionsKt___Collections.kt:288)
       at com.onesignal.common.modeling.ModelStore.add(ModelStore.kt:41)
       at com.onesignal.common.modeling.IModelStore$DefaultImpls.add$default(IModelStore.kt:37)
       at com.onesignal.core.internal.operations.impl.OperationRepo.internalEnqueue(OperationRepo.kt:76)
       at com.onesignal.core.internal.operations.impl.OperationRepo.enqueue(OperationRepo.kt:60)
       at com.onesignal.core.internal.operations.IOperationRepo$DefaultImpls.enqueue$default(IOperationRepo.kt:16)
       at com.onesignal.core.internal.operations.listeners.SingletonModelStoreListener.onModelUpdated(SingletonModelStoreListener.kt:51)
       at com.onesignal.common.modeling.SingletonModelStore$onModelUpdated$1.invoke(SingletonModelStore.kt:50)
       at com.onesignal.common.modeling.SingletonModelStore$onModelUpdated$1.invoke(SingletonModelStore.kt:50)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.SingletonModelStore.onModelUpdated(SingletonModelStore.kt:50)
       at com.onesignal.common.modeling.ModelStore$onChanged$1.invoke(ModelStore.kt:74)
       at com.onesignal.common.modeling.ModelStore$onChanged$1.invoke(ModelStore.kt:74)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.ModelStore.onChanged(ModelStore.kt:74)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.Model.notifyChanged(Model.kt:296)
       at com.onesignal.common.modeling.Model.notifyChanged(Model.kt:301)
       at com.onesignal.common.modeling.Model.setOptAnyProperty(Model.kt:201)
       at com.onesignal.common.modeling.Model.setOptAnyProperty$default(Model.kt:188)
       at com.onesignal.common.modeling.MapModel.put(MapModel.kt:47)
       at com.onesignal.common.modeling.MapModel.remove(MapModel.kt:8)
       at com.onesignal.common.modeling.MapModel.put(MapModel.kt:8)
       at com.onesignal.user.internal.UserManager.addTag(UserManager.kt:173)
```

```
Fatal Exception: java.util.ConcurrentModificationException:
       at java.util.ArrayList$Itr.next(ArrayList.java:860)
       at com.onesignal.common.modeling.ModelStore.persist(ModelStore.kt:140)
       at com.onesignal.common.modeling.ModelStore.onChanged(ModelStore.kt:72)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.Model.notifyChanged(Model.kt:296)
       at com.onesignal.common.modeling.Model.setOptAnyProperty(Model.kt:201)
       at com.onesignal.user.internal.subscriptions.SubscriptionModel.setStatus(SubscriptionModel.kt:128)
       at com.onesignal.user.internal.subscriptions.impl.SubscriptionManager.addOrUpdatePushSubscription(SubscriptionManager.kt:71)
       at com.onesignal.notifications.internal.listeners.DeviceRegistrationListener$retrievePushTokenAndUpdateSubscription$1.invokeSuspend(DeviceRegistrationListener.kt:74)
       at com.onesignal.notifications.internal.listeners.DeviceRegistrationListener$retrievePushTokenAndUpdateSubscription$1.invoke(DeviceRegistrationListener.kt:1)
       at com.onesignal.notifications.internal.listeners.DeviceRegistrationListener$retrievePushTokenAndUpdateSubscription$1.invoke(DeviceRegistrationListener.kt:2)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1$1.invokeSuspend(ThreadUtils.kt:61)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
       at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
       at kotlinx.coroutines.BlockingCoroutine.joinBlocking(BlockingCoroutine.java:85)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(BuildersKt__Builders.kt:59)
       at kotlinx.coroutines.BuildersKt.runBlocking(Builders.kt:1)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(BuildersKt__Builders.kt:38)
       at kotlinx.coroutines.BuildersKt.runBlocking$default(Builders.kt:1)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:60)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:59)
       at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30)
```

```
Fatal Exception: java.util.ConcurrentModificationException:
       at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1029)
       at java.util.ArrayList$Itr.next(ArrayList.java:982)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:43)
       at com.onesignal.core.internal.application.impl.ApplicationService.setCurrent(ApplicationService.kt:50)
       at com.onesignal.core.internal.application.impl.ApplicationService.onActivityStarted(ApplicationService.kt:145)
       at android.app.Application.dispatchActivityStarted(Application.java:401)
       at android.app.Activity.dispatchActivityStarted(Activity.java:1452)
       at android.app.Activity.onStart(Activity.java:1991)
       at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1582)
       at android.app.Activity.performStart(Activity.java:8628)
       at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3807)
       at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:225)
       at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:205)
       at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:177)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:98)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2443)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:205)
       at android.os.Looper.loop(Looper.java:294)
       at android.app.ActivityThread.main(ActivityThread.java:8177)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
```

```
Fatal Exception: java.util.ConcurrentModificationException:
       at java.util.ArrayList$Itr.next(ArrayList.java:860)
       at kotlin.collections.CollectionsKt___CollectionsKt.firstOrNull(CollectionsKt___Collections.kt:288)
       at com.onesignal.common.modeling.ModelStore.add(ModelStore.kt:41)
       at com.onesignal.common.modeling.IModelStore$DefaultImpls.add$default(IModelStore.kt:37)
       at com.onesignal.core.internal.operations.impl.OperationRepo.internalEnqueue(OperationRepo.kt:76)
       at com.onesignal.core.internal.operations.impl.OperationRepo.enqueue(OperationRepo.kt:60)
       at com.onesignal.core.internal.operations.IOperationRepo$DefaultImpls.enqueue$default(IOperationRepo.kt:16)
       at com.onesignal.core.internal.operations.listeners.ModelStoreListener.onModelUpdated(ModelStoreListener.kt:49)
       at com.onesignal.common.modeling.ModelStore$onChanged$1.invoke(ModelStore.kt:74)
       at com.onesignal.common.modeling.ModelStore$onChanged$1.invoke(ModelStore.kt:74)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.ModelStore.onChanged(ModelStore.kt:74)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.modeling.Model$notifyChanged$1.invoke(Model.kt:296)
       at com.onesignal.common.events.EventProducer.fire(EventProducer.kt:44)
       at com.onesignal.common.modeling.Model.notifyChanged(Model.kt:296)
       at com.onesignal.common.modeling.Model.setOptAnyProperty(Model.kt:201)
       at com.onesignal.user.internal.subscriptions.SubscriptionModel.setStatus(SubscriptionModel.kt:128)
       at com.onesignal.user.internal.subscriptions.impl.SubscriptionManager.addOrUpdatePushSubscription(SubscriptionManager.kt:71)
       at com.onesignal.notifications.internal.listeners.DeviceRegistrationListener$retrievePushTokenAndUpdateSubscription$1.invokeSuspend(DeviceRegistrationListener.kt:74)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:32)
       at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
       at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
       at kotlinx.coroutines.BlockingCoroutine.joinBlocking(BlockingCoroutine.java:85)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(BuildersKt__Builders.kt:59)
       at kotlinx.coroutines.BuildersKt.runBlocking(Builders.kt:1)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(BuildersKt__Builders.kt:38)
       at kotlinx.coroutines.BuildersKt.runBlocking$default(Builders.kt:1)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:60)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:59)
       at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30)
```

```
Fatal Exception: java.util.ConcurrentModificationException:
       at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:757)
       at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:790)
       at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:788)
       at com.onesignal.common.modeling.Model.toJSON(Model.kt:312)
       at com.onesignal.common.modeling.ModelStore.persist(ModelStore.kt:141)
       at com.onesignal.common.modeling.SingletonModelStore.replace(SingletonModelStore.kt:36)
       at com.onesignal.core.internal.config.impl.ConfigModelStoreListener$fetchParams$1.invokeSuspend(ConfigModelStoreListener.kt:101)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:32)
       at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
       at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
       at kotlinx.coroutines.BlockingCoroutine.joinBlocking(BlockingCoroutine.java:85)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(BuildersKt__Builders.kt:59)
       at kotlinx.coroutines.BuildersKt.runBlocking(Builders.kt:1)
       at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(BuildersKt__Builders.kt:38)
       at kotlinx.coroutines.BuildersKt.runBlocking$default(Builders.kt:1)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:60)
       at com.onesignal.common.threading.ThreadUtilsKt$suspendifyOnThread$1.invoke(ThreadUtils.kt:59)
       at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30)
```

The only one that we can reproduce in a test device is the one related to the `EventProducer.fire` call.
So we tried to update the sdk to the latest version 5.0.4 in order to check if the `EventProducer.fire` crash is resolved, but it occurs again every time. 

### Steps to reproduce?

```Markdown
We cannot reproduce these issue, only the EventProducer.fire calls with our integration (I cannot share it here)
```

### What did you expect to happen?

No crash on user device

### OneSignal Android SDK version

5.0.0

### Android version

13, 12, 11, 10, 9, 8

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jennantilla** [comment]
Hi @sarbyn thank you for reaching out! I know you mention you cannot share your integration but can you share any additional details about the steps you used for reproducing a crash with `EventProducer.fire` in 5.0.4? Have you upgraded your production app to 5.0.4 yet? If so, have you seen any of the other non-reproducible crashes? Thanks!

**Comment by sarbyn** [comment]
The production app is not yet upgraded to 5.0.4, so we cannot check if other crashes are fixed in production.

In our implementation, at startup, the app could send 3 different user Tag at boot time, like this pseudocode

```
// Before this call we 
// * Step 1: check the system notification permission
//  * Step 2: e call one signal login
// this functionAtStartup is our step 3
fun functionAtStartup() {
    OneSignal.User.addTag("KEY1", "alpha")
    OneSignal.User.addTag("KEY2", "beta")
    OneSignal.User.addTag("KEY3", "gamma")
}
```

If the user DENY, on android 13 and later, the notification permission in Step 1, when functionAtStartup is called, the app crashes

**Comment by epegasus** [comment]
I'm using **v5.0.4**, and I'm also facing this issue in this version.

## Devices            honor, samsung
## Android version    13


Exception java.util.ConcurrentModificationException:
  at java.util.ArrayList$Itr.checkForComodification (ArrayList.java:1029)
  at java.util.ArrayList$Itr.next (ArrayList.java:982)
  at com.onesignal.common.events.EventProducer.fire (EventProducer.kt:49)
  at com.onesignal.core.internal.application.impl.ApplicationService.setCurrent (ApplicationService.kt:50)
  at com.onesignal.core.internal.application.impl.ApplicationService.onActivityStarted (ApplicationService.kt:150)
  at android.app.Application.dispatchActivityStarted (Application.java:408)
  at android.app.Activity.dispatchActivityStarted (Activity.java:1469)
  at android.app.Activity.onStart (Activity.java:2001)
  at android.app.Instrumentation.callActivityOnStart (Instrumentation.java:1525)
  at android.app.Activity.performStart (Activity.java:8706)
  at android.app.ActivityThread.handleStartActivity (ActivityThread.java:4414)
  at android.app.servertransaction.TransactionExecutor.performLifecycleSequence (TransactionExecutor.java:235)
  at android.app.servertransaction.TransactionExecutor.cycleToPath (TransactionExecutor.java:215)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:187)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:105)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2838)
  at android.os.Handler.dispatchMessage (Handler.java:114)
  at android.os.Looper.loopOnce (Looper.java:206)
  at android.os.Looper.loop (Looper.java:296)
  at android.app.ActivityThread.main (ActivityThread.java:9159)
  at java.lang.reflect.Method.invoke
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:591)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1018)

**Comment by KChernenko** [comment]
I'm also facing this issue in the v5.0.4.
The similar devices, the similar stacktrace posted above.

**Comment by jennantilla** [comment]
Hello, everyone! Thank you for your patience on this issue. A new fix to address these additional ConcurrentModificationException was released in [5.1.1](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/5.1.1). Please update to this version and let us know if you are still having any issues.
